rfd/0107-desktop-access-performance-research.md

```diff
@@ -0,0 +1,95 @@
+---
+authors: Łukasz Kozłowski (lukasz.kozlowski@goteleport.com)
+state: draft
+---
+
+# RFD 107 - Desktop Access Performance
+
+## Required Approvers
+Engineering: @zmb3 && @ibeckermayer
+
+## What
+
+This RFD explores possible solutions to the current performance problems in the destkop
+access module.
+
+## Why
+
+We want to improve user experiance when using the desktop access module. Currently the main issue
+is insufficent performance of the video rendering which results in choppier experience or noticeable stuttering. Each frame in the RDP protocol consists of (in most cases) 64x64 pixel bitmaps that are assembeld into the full frame. All of these bitmaps needs to be send by the server, received and processed by the client. There are different ways in which we can reduce time to render a frame.
+One of the ways is to reduce number of bitmaps that we need to send over the wire. The other way is to reduce the time to process individual bitmap. In this document we'll discuss some solutions that might fix these problems.
+
+
+### RDP Bitmap caching
+
+One of the ways to improve the performance of Teleport Desktop Access is to reduce the amount of data sent between the proxy and the web browser and to reduce the number of messages with bitmap data that needs processing. One RDP protocol extensions address the problem of reducing latency and bandwidth usage. It's called [[MS-RDPEGDI] Remote Desktop Protocol: Graphics Device Interface (GDI) Acceleration Extensions](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/745f2eee-d110-464c-8aca-06fc1814f6ad)
```

Comment:
@ibeckermayer also pointed out [MS-RDPEGFX](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegfx/da5c75f9-cd99-450c-98c4-014a496942b0)

How does MS-RDPEGFX compare to MS_RDPEGDI?

Reply:
First of all, these two extensions are unrelated to each other. 

The "Remote Desktop Protocol: Graphics Device Interface (GDI)
Acceleration Extensions [MS-RDPEGDI]" was released on 2/22/2007 and is available starting from the RDP 6.0.

The  "Remote Desktop Protocol: Graphics Pipeline Extension [MS-RDPEGFX]" was released on 12/16/2011 and is available starting from the RDP 8.0.

Both extensions improve performance and reduce the number of operations the client needs to do to render a full frame. They define and use different primitives to achieve this goal.

The [MS-RDPEGDI] extension specifies drawing operations, like combining bitmap patterns using raster operations and painting glyphs, rectangles, lines, and ellipses. It also allows transferring portions of the screen area from one point to another. It also defines bitmap caches and operations to store and load bitmap data from these caches.

On the other hand, the [MS-RDPEGFX] extension introduces surfaces and operations that act upon them. There is also a graphics output buffer which is an end-user visible bitmap. 
It works more or less as follows: 
- create surfaces on the client
- transfer encoded bitmap data from the server to a client-side destination surface
- map surfaces to the graphics output buffer
- render frame

There are also operations that allow copying bitmap data from one surface to another. This extension also defines bitmap caches and operations that can store and load bitmap data in these caches. It also uses different and newer codecs to encode bitmap information e.g:
- RemoteFX
- ClearCodec
- MPEG-4 AVC/H.264

The client using this extension might be more performant, but it defines a new way of rendring bitmap data. It also requires support of a lot more operations and a newer version of the RDP compared to only using caches operations from the [MS-RDPEGDI].

Reply:
> It also uses different and newer codecs to encode bitmap information e.g:
>
> - RemoteFX
> - ClearCodec
> - MPEG-4 AVC/H.264

Technically RemoteFX is a part of [MS-RDPEGDI]: see [codecGUID](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/86507fed-a0ee-4242-b802-237534a8f65e)

This slide is helpful
<img width="1247" alt="image" src="https://user-images.githubusercontent.com/13578537/221306753-24971be0-3f3e-48b3-9bc8-e94360f63ae0.png">

([source](https://interopevents.blob.core.windows.net/uploads/PDFs/2019/RDPIO/RDP%20IO%20Lab%20-%20RDP%20Update%202019%20-Denis%20Gundarev.pdf))